joe/joe.py

```diff
@@ -40,40 +42,42 @@ def _get_data_dir(path):
 
 def _walk_gitignores():
     '''Recurse over the data directory and return all .gitignore file names'''
-    l = []
-    for root, subFolders, files in os.walk(DATA_DIR):
-        l += [f.replace('.gitignore', '') \
-                for f in files if f.endswith('.gitignore')]
-    return sorted(l)
+    gitignores = []
+    for _root, _sub_folders, files in os.walk(DATA_DIR):
+        gitignores += [f.replace('.gitignore', '')
+                       for f in files if f.endswith('.gitignore')]
+    return sorted(gitignores)
 
 
 # Where all gitignore files are
 DATA_DIR = _get_data_dir('*.gitignore')
 # Load up names for all gitignore files
 GITIGNORE_RAW = _walk_gitignores()
-GITIGNORE = [f.lower() for f in GITIGNORE_RAW]
+GITIGNORE = [filename.lower() for filename in GITIGNORE_RAW]
 
 
-def _print_filenames():
+def _get_filenames():
     '''List all available .gitignore files.'''
-    print ', '.join(GITIGNORE)
+    return ', '.join(GITIGNORE)
 
 
 def _handle_gitignores(names):
-    '''Generates and sends the gitignore contents to stdout.'''
-    exit = False
+    '''Generates and returns the gitignore contents.'''
     output = '#### joe made this: https://goel.io/joe\n'
+    failed = []
     for name in names:
         try:
             raw_name = GITIGNORE_RAW[GITIGNORE.index(name.lower())]
+            output += _fetch_gitignore(raw_name)
         except ValueError:
-            print ('Uh oh! Seems like joe doesn\'t know what %s is.\n'
-                   'Try running `joe ls` to see list of available gitignore '
-                   'files.') % name
-            exit = True
-            break
-        output += _fetch_gitignore(raw_name)
-    print output
+            failed.append(name)
+    if failures:
```

Comment:
Oops! You're totally correct on the typo. However, for the empty array, that is generally checked in the manner `if arr`, because all non-empty arrays are not falsey. It's considered standard to do a boolean check.